File: cpp/set-active-profile/src/SetActiveProfile.cpp
/*
 * (c) 2018-2021 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <stdio.h>
#include <cstdlib>
#include <iostream>
#include "CrossPlatform.h"

#ifdef _WIN32
    #define HOMEVAR "USERPROFILE"
#else 
    #define HOMEVAR "HOME"
#endif

int main(int argc, char* argv[]) {

    int nErrorCode;
    std::string persistorPath = "../../sample-data/persistors/sample-persistor.pt";
    std::string profileId = "ABcd.1.48sdf0-cs80-5802-sd80-d8s0df80sdfj";

    // read SDK path to use to load the crypto libs from environment variable
    char* cSdkPath = std::getenv("IONIC_SDK_PATH");
    if (cSdkPath == NULL) {
        std::cerr << "[!] Please provide the SDK path as env variable: IONIC_SDK_PATH" << std::endl;
        exit(1);
    }
    std::string sdkPath = std::string(cSdkPath);
    std::string cryptoPath = sdkPath + "/ISAgentSDKCpp/Lib/" + OS + "/Release/" + ARCH;
    ISCrypto::setCryptoSharedLibraryCustomDirectory(cryptoPath);

    // initialize agent with plaintext persistor
    ISAgent agent;
    ISAgentDeviceProfilePersistorPlaintext persistor;
    persistor.setFilePath(persistorPath);
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Error initializing agent: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
    }

    // list all available profiles
    const std::vector<ISAgentDeviceProfile> profiles = agent.getAllProfiles();
    std::cout << "ALL_PROFILES:" << std::endl;
    for (int i=0; i < profiles.size(); i++) {
        const ISAgentDeviceProfile profile = profiles.at(i);
        std::cout << profile.getDeviceId() << std::endl;
    }

    // set new active profile
    bool bSuccess = agent.setActiveProfile(profileId);
    if (!bSuccess) {
        std::cerr << "Failed to set active profile" << std::endl;
        exit(2);
    }

    // display new active profile
    ISAgentDeviceProfile activeProfile = agent.getActiveProfile();
    std::cout << std::endl << "ACTIVE_PROFILE:" << std::endl;
    std::cout << activeProfile.getDeviceId() << std::endl;
}

File: java/set-active-profile/src/main/java/com/ionic/samples/SetActiveProfile.java
/*
 * (c) 2018 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 * 
 * Developed with Ionic Java SDK 2.1.0
 */

package com.ionic.samples;

import com.ionic.sdk.device.profile.persistor.DeviceProfilePersistorPlainText;
import com.ionic.sdk.device.profile.DeviceProfile;
import com.ionic.sdk.agent.Agent;
import com.ionic.sdk.error.IonicException;

public class SetActiveProfile
{
    public static void main(String[] args)
    {
        String profileId = "ABcd.1.48sdf0-cs80-5802-sd80-d8s0df80sdfj";

        // initialize agent
        Agent agent = new Agent();
        try {
            String persistorPath = "../../sample-data/persistors/sample-persistor.pt";
            DeviceProfilePersistorPlainText persistor = new DeviceProfilePersistorPlainText(persistorPath);
            agent.initialize(persistor);
        } catch(IonicException e) {
            System.out.println(e.getMessage());
            System.exit(1);
        }
        
        // verify there is at least one profile
        if (agent.getAllProfiles().size == 0) {
            System.println("No profiles found in persistor");
            System.exit(1);
        }
        
        // list all profiles
        System.out.println("All Profiles:");
        for (DeviceProfile profile : agent.getAllProfiles()) {
            System.out.println("---");
            System.out.println(profile.getDeviceId());
        }

        // 

        /*
        // set active profile
        agent.setActiveProfile(profileId);

        // display agent active profile
        DeviceProfile profile = agent.getActiveProfile();
        System.out.println("Id       : " + profile.getDeviceId());
        System.out.println("Name     : " + profile.getName());
        System.out.println("Keyspace : " + profile.getKeySpace());
        System.out.println("ApiUrl   : " + profile.getServer());
        */
    }
}
